### PR TITLE
Refactor thread management and shutdown

### PR DESCRIPTION
**Pull request by hlinnaka**
See commit messages.

There's been many previous attempts at these things, see https://github.com/zenithdb/zenith/pull/787, https://github.com/zenithdb/zenith/pull/1088, https://github.com/zenithdb/zenith/pull/1069. This is yet another attempt. I tried to make things more clear by introducing a global registry of threads, and some common facilities for requesting thread shutdown.

### CODE REVIEW

File: pageserver/src/walreceiver.rs [hunk 1]
```diff
@@ -156,25 +107,18 @@ fn get_wal_producer_connstr(timelineid: ZTimelineId) -> String {
 //
 fn thread_main(
     conf: &'static PageServerConf,
-    timelineid: ZTimelineId,
     tenantid: ZTenantId,
-    interrupt_receiver: oneshot::Receiver<()>,
-) {
+    timelineid: ZTimelineId,
+) -> Result<()> {
```
Comment (SomeoneToIgnore) [inline pageserver/src/walreceiver.rs:112]: NIT: you don't really need to return the result here, and the closure that invokes this method can just wrap it in Ok(launch_wal_receiver(...)).

Might be more than one place that has this.

File: pageserver/src/thread_mgr.rs [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    // Thread that connects to a safekeeper to fetch WAL for one timeline.
+    WalReceiver,
+
+    // Thread that handles checkpointing of all timelines for a tenant.
+    Checkpointer,
+
+    // Thread that handles GC of a tenant
+    GarbageCollector,
+
+    // Thread that checks status of remote storage. Shared by all tenants.
```
Comment (SomeoneToIgnore) [inline pageserver/src/thread_mgr.rs:81]:
```suggestion
    // Thread for synchronizing pageserver relish data with the remote storage. Shared by all tenants.
```

Also, as an idea to implement later:
> Shared by all tenants

could be enforced on a type level, so that we could not create such threads wrongfully.

File: pageserver/src/thread_mgr.rs [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+/// This wrapper function runs in a newly-spawned thread. It initialies the thread-local
+/// variables and calls the payload function
+fn thread_wrapper<F, E>(
+    thread_id: u64,
+    thread: Arc<PageServerThread>,
+    shutdown_rx: watch::Receiver<()>,
+    f: F,
+) where
+    F: FnOnce() -> Result<(), E>,
+    F: Send + 'static,
```
Comment (SomeoneToIgnore) [inline pageserver/src/thread_mgr.rs:177]: NIT: there is a couple of places, where you can merge the trait bounds into one line.

File: pageserver/src/thread_mgr.rs [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+//! Depending on what thread panics, we might want to kill the whole server, or
+//! only a single tenant or timeline.
+//!
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::panic;
+use std::panic::AssertUnwindSafe;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
```
Comment (SomeoneToIgnore) [inline pageserver/src/thread_mgr.rs:20]: > `Mutex`

Do you think we can reuse parking lot here too?

File: pageserver/src/thread_mgr.rs [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+///   is_thread_running(None, Some(timelineid), None)
+///
+pub fn is_thread_running(
+    kind: Option<ThreadKind>,
+    tenant_id: Option<ZTenantId>,
+    timeline_id: Option<ZTimelineId>,
+) -> bool {
+    let threads = THREADS.lock().unwrap();
+    for thread in threads.values() {
+        if kind.is_none() || thread.kind == kind.unwrap() {
```
Comment (SomeoneToIgnore) [inline pageserver/src/thread_mgr.rs:214]: NIT: That unused function is a bit odd IMO, the `unwrap` could be rewritten in more comprehensible way, also, `is_none` path does not really require any locking, which might be nice too.
IMO, `match` is enough to handle those in a readable manner.

Similar idea for the shutdown function.

File: pageserver/src/thread_mgr.rs [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        inner_guard.tenant_id = tenant_id;
+        inner_guard.timeline_id = timeline_id;
+    })
+}
+
+/// A Future that can be used to check if the current thread has been requested to
+/// shut down.
+pub async fn shutdown_requested() {
+    let _ = SHUTDOWN_RX
+        .with(|rx| rx.borrow().as_ref().unwrap().clone())
```
Comment (SomeoneToIgnore) [inline pageserver/src/thread_mgr.rs:280]: I'm not fond of `unwrap()` in general, so I wonder, would it make sense to either
* avoid panicking (log and error and move on?)
* panic with `context` that explains what invariant did we break so hard that we panic

Sure verbose things like `lock().unwrap()` are exception, but `Option::unwrap()` is already a nice thing to document IMO.

**Review by SomeoneToIgnore — COMMENTED** [review]
I like how the PR touches the same ideas I've had (a global place to shut down everything) and an async approach to that.

Not sure I understand all the locking interleaves, so somebody else is more than welcome to verify that threads aren't interfering the wrong way.

**Comment by SomeoneToIgnore** [comment]
Also fixes https://github.com/zenithdb/zenith/issues/1036

File: pageserver/src/thread_mgr.rs (hunk 4 above)
Comment (hlinnaka) [reply pageserver/src/thread_mgr.rs:20]: Sure, although std::sync::Mutex is still what most places use. I'm in favor of switching the whole codebase to parking_lot, but I'd prefer to do it as one separate commit.

File: pageserver/src/thread_mgr.rs (hunk 2 above)
Comment (hlinnaka) [reply pageserver/src/thread_mgr.rs:81]: > Also, as an idea to implement later:
> 
> > Shared by all tenants
> 
> could be enforced on a type level, so that we could not create such threads wrongfully.

Yeah, I considered having 'timelineid' and 'tenantid' part of this Enum, for those thread kinds that are associated with timelines or tenants. But that's a bit strange for PageRequestHandlers which are initially not associated with timelines, but become associated later.

Now that I look more closely, I actually forgot to call `associate_thread` anywhere. So PageRequestHandlers are never associated with a timeline. That means that any threads that are currently handling requests for a timeline are not killed, when the timeline is shut down. Maybe that's OK and I was just trying to be too cute with this. I think I'll remove `associate_thread` since it's dead code, and leave that for a later patch if we want to do it.

Comment (SomeoneToIgnore) [reply pageserver/src/thread_mgr.rs:81]: In general, I'd be very cautious when adding `pub` modifiers to such methods/functions, as it turns off rustc's warnings about unused.

File: pageserver/src/thread_mgr.rs (hunk 5 above)
Comment (hlinnaka) [reply pageserver/src/thread_mgr.rs:214]: > NIT: That unused function is a bit odd

Ah, didn't realize it was unused. I had a use case for it in mind at some point, but don't remember what it was. Removed.

> IMO, the unwrap could be rewritten in more comprehensible way, also, is_none path does not really require any locking, which might be nice too.
IMO, match is enough to handle those in a readable manner.

Can you help me write that with `match`?

File: pageserver/src/thread_mgr.rs (hunk 6 above)
Comment (hlinnaka) [reply pageserver/src/thread_mgr.rs:280]: Changed it to use `Option::expect`, so you get a panic with a better message.

File: pageserver/src/walreceiver.rs (hunk 1 above)
Comment (hlinnaka) [reply pageserver/src/walreceiver.rs:112]: I tried that now, but got:
```
  --> pageserver/src/walreceiver.rs:71:13
   |
71 |             thread_mgr::spawn(
   |             ^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `E` declared on the function `spawn`
```
This fixes it:
```
            thread_mgr::spawn(
                ThreadKind::WalReceiver,
                Some(tenantid),
                Some(timelineid),
                "WAL receiver thread",
                move || {
                    IS_WAL_RECEIVER.with(|c| c.set(true));
                    thread_main(conf, tenantid, timelineid);
                    Ok::<(), anyhow::Error>(())
                },
            )?;
```
But that ` Ok::<(), anyhow::Error>(())` looks pretty complicated, so I'm inclined to leave it as it is.

File: pageserver/src/thread_mgr.rs (hunk 5 above)
Comment (SomeoneToIgnore) [reply pageserver/src/thread_mgr.rs]: Hm, either I've got wrong, or there were two locks here in some verison, now that's gone and not that much can be really done.

I'd replaced `thread.kind == kind.unwrap()` with `kind == Some(thread.kind)`, but that's not necessary.

File: pageserver/src/repository.rs [hunk 7]
```diff
@@ -19,8 +19,6 @@ pub type BlockNumber = u32;
 /// A repository corresponds to one .zenith directory. One repository holds multiple
 /// timelines, forked off from the same initial call to 'initdb'.
 pub trait Repository: Send + Sync {
-    fn shutdown(&self) -> Result<()>;
```
Comment (LizardWizzard) [inline pageserver/src/repository.rs:22]: Do not get it who is responsible for performing checkpoint when pageserver is shutting down. I see that shutdown_timeline code migrated to put_timelines_into_tenant but I do not see where we trigger this code when pageserver is stopped

**Review by LizardWizzard — COMMENTED** [review]
Nice!

I've left some mostly minor comments. 

I think shutdown topic deserves it's own documentation page (thread management, who is responsible for flushing all needed data to disk etc). This approach with thread local flags/ async watchers may be a bit hard to follow at first. So I think documentation can help with that

File: pageserver/src/layered_repository.rs [hunk 8]
```diff
@@ -357,30 +340,6 @@ impl Repository for LayeredRepository {
...
 
-fn shutdown_timeline(
-    tenant_id: ZTenantId,
-    timeline_id: ZTimelineId,
-    timeline: &LayeredTimelineEntry,
-) -> Result<(), anyhow::Error> {
-    match timeline {
-        LayeredTimelineEntry::Local(timeline) => {
-            timeline
-                .upload_relishes
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:369]: I did some more thinking about this. The possible states of a timeline are:
```
/// A state of the timeline synchronization with the remote storage.
/// Contains `disk_consistent_lsn` of the corresponding remote timeline (latest checkpoint's disk_consistent_lsn).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
pub enum TimelineSyncState {
    /// No further downloads from the remote storage are needed.
    /// The timeline state is up-to-date or ahead of the remote storage one,
    /// ready to be used in any pageserver operation.
    Ready(Lsn),
    /// Timeline is scheduled for downloading, but its current local state is not up to date with the remote storage.
    /// The timeline is not ready to be used in any pageserver operations, otherwise it might diverge its local state from the remote version,
    /// making it impossible to sync it further.
    AwaitsDownload(Lsn),
    /// Timeline was not in the pageserver's local working directory, but was found on the remote storage, ready to be downloaded.
    /// Cannot be used in any pageserver operations due to complete absence locally.
    CloudOnly(Lsn),
    /// Timeline was evicted from the pageserver's local working directory due to conflicting remote and local states or too many errors during the synchronization.
    /// Such timelines cannot have their state synchronized further and may not have the data about remote timeline's disk_consistent_lsn, since eviction may happen
    /// due to errors before the remote timeline contents is known.
    Evicted(Option<Lsn>),
}
```

If we're switching to Ready state, then we don't do anything here, before or after this PR. In all the other states, does it even make sense to perform a final checkpoint before changing the state? I don't think it does. In all of those other states, we're not going to use the local data for anything anymore, so what's the point of the final checkpoint? Does that make sense, @SomeoneToIgnore? So I changed this so that we don't perform the final checkpoint.

I also refactored the code in `set_timeline_states`, moving the code to create the LayeredRepository object so that the `Tenant.repo` field can no longer be None. The None was just a transient state between initializing the Tenant and its LayeredRepository, but now that transient state doesn't happen anymore.

File: pageserver/src/repository.rs (hunk 7 above)
Comment (hlinnaka) [reply pageserver/src/repository.rs:22]: `shutdown_all_tenants`

Comment (LizardWizzard) [reply pageserver/src/repository.rs:22]: Got it, thanks!

File: walkeeper/src/bin/safekeeper.rs [hunk 9]
```diff
@@ -197,7 +197,12 @@ fn start_safekeeper(conf: SafeKeeperConf) -> Result<()> {
             .spawn(|| {
                 // TODO authentication
                 let router = http::make_router(conf_);
-                endpoint::serve_thread_main(router, http_listener).unwrap();
+                endpoint::serve_thread_main(
+                    router,
+                    http_listener,
+                    std::future::pending(), // never shut down
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply walkeeper/src/bin/safekeeper.rs:203]: Safekeeper's shutdown sequence is a lot more straightforward at the moment:
```
    // NOTE: we still have to handle signals like SIGQUIT to prevent coredumps
    signals.handle(|signal| {
        // TODO: implement graceful shutdown with joining threads etc
        info!(
            "Got {}. Terminating in immediate shutdown mode",
            signal.name()
        );
        std::process::exit(111);
    })
```

:-). It has worked for us so far.

File: pageserver/src/thread_mgr.rs [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+        } else {
+            // The thread had not even fully started yet. Or it was shut down
+            // concurrently and alrady exited
+        }
+    }
+}
+
+/// A Future that can be used to check if the current thread has been requested to
+/// shut down.
+pub async fn shutdown_requested() {
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply pageserver/src/thread_mgr.rs]: +1 renamed to `shutdown_watcher`

File: pageserver/src/layered_repository.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply pageserver/src/layered_repository.rs:369]: Yes, I think there's no need to have the checkpoint indeed, but would be nice to set the upload flag to `false` if we don't do that yet.

**Review by SomeoneToIgnore — APPROVED** [review]

**Comment by hlinnaka** [comment]
Thanks for the review!

> I think shutdown topic deserves it's own documentation page (thread management, who is responsible for flushing all needed data to disk etc). This approach with thread local flags/ async watchers may be a bit hard to follow at first. So I think documentation can help with that

I tried to improve the comments, I hope that helps

File: pageserver/src/thread_mgr.rs [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+        }
+    };
+    *jh_guard = Some(join_handle);
+    drop(jh_guard);
+
+    // The thread is now running. Nothing more to do here
+    Ok(())
+}
+
+/// This wrapper function runs in a newly-spawned thread. It initialies the thread-local
```
Comment (LizardWizzard) [inline pageserver/src/thread_mgr.rs:181]: typo: initialies -> initialises

File: pageserver/src/thread_mgr.rs [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+
+/// Is there a thread running that matches the criteria
+
+/// Signal and wait for threads to shut down.
+///
+///
+/// The arguments are used to select the threads to kill. Any None arguments are
+/// ignored. For example, to shut down all WalReceiver threads:
+///
+///   is_thread_running(Some(ThreadKind::WalReceiver), None, None)
```
Comment (LizardWizzard) [inline pageserver/src/thread_mgr.rs:216]: Probably shutdown_threads meant to be used here

File: walkeeper/src/bin/safekeeper.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply walkeeper/src/bin/safekeeper.rs:203]: 😃

**Review by LizardWizzard — APPROVED** [review]
LGTM

(it seems that there are some accidentally included postgres changes)

**Comment by hlinnaka** [comment]
Pushed, thanks for the reviews!

There's more we could do here, like shutdown the whole server or a tenant on a panic, depending on which thread panic'd, and shutting down a single tenant is still not implemented. But this is a step in the right direction, and fixes some live issues we've been having.
